File: apps/vsc-plugin/src/services/agent-detection/detectors/cursor.detector.ts
import { whichBinary } from '../checks/binary';
import { dirExists, expandHome } from '../checks/config-dir';
import type { AgentDetector, DetectionContext, DetectionResult } from '../types';

/**
 * Detects Cursor's headless CLI agent. Two signals, first hit wins:
 *
 *   1. The `cursor-agent` binary on PATH.
 *   2. The `~/.cursor/` config directory (created by Cursor IDE on
 *      first launch; also produced by the `cursor-agent login` flow).
 *
 * Cursor itself is a VS Code fork that does NOT publish a marketplace
 * extension into upstream VS Code, so there's no extension probe — the
 * binary / config-dir cover both the standalone-CLI and inside-Cursor
 * install paths.
 *
 * Every branch returns `isTerminalAgent: true` so the mobile knows to
 * dispatch runtime commands (start_task / select_option / …) to
 * codeam-cli's pluginId rather than this plugin.
 */
export class CursorDetector implements AgentDetector {
  readonly id = 'cursor';
  readonly name = 'Cursor Agent';
  readonly icon = 'cursor';

  private static readonly BINARY_NAME = 'cursor-agent';
  private static readonly CONFIG_DIR = '~/.cursor';

  async detect(ctx: DetectionContext): Promise<DetectionResult | null> {
    const binPath = await whichBinary(CursorDetector.BINARY_NAME);
    if (binPath) {
      ctx.log.appendLine(`[CursorDetector] binary at ${binPath}`);
      return {
        installed: true,
        extensionId: `__binary__:${CursorDetector.BINARY_NAME}`,
        isTerminalAgent: true,
        via: 'binary',
      };
    }

    if (await dirExists(expandHome(CursorDetector.CONFIG_DIR))) {
      ctx.log.appendLine(`[CursorDetector] config dir ${CursorDetector.CONFIG_DIR} present`);
      return {
        installed: true,
        extensionId: `__config__:${CursorDetector.BINARY_NAME}`,
        isTerminalAgent: true,
        via: 'config-dir',
      };
    }

    return null;
  }
}

File: apps/jetbrains-plugin/src/main/kotlin/com/windsurf/controller/services/detection/detectors/CursorDetector.kt
package com.windsurf.controller.services.detection.detectors

import com.windsurf.controller.services.detection.AgentDetector
import com.windsurf.controller.services.detection.DetectionContext
import com.windsurf.controller.services.detection.DetectionResult
import com.windsurf.controller.services.detection.checks.dirExists
import com.windsurf.controller.services.detection.checks.expandHome
import com.windsurf.controller.services.detection.checks.whichBinary

/**
 * Cursor's headless CLI agent — terminal agent owned by codeam-cli.
 * Cursor is a VS Code fork that doesn't ship a JetBrains plugin, so
 * detection is limited to the `cursor-agent` binary on PATH and the
 * `~/.cursor` config directory (created on first IDE launch / agent
 * login).
 */
class CursorDetector : AgentDetector {
    override val id = "cursor"
    override val name = "Cursor Agent"
    override val icon = "cursor"

    private val binaryName = "cursor-agent"
    private val configDir = "~/.cursor"

    override suspend fun detect(ctx: DetectionContext): List<DetectionResult> {
        val installed = whichBinary(binaryName) != null || dirExists(expandHome(configDir))
        if (!installed) return emptyList()
        ctx.logger.info("[CursorDetector] detected")
        return listOf(
            DetectionResult(
                id = id,
                name = name,
                icon = icon,
                pluginId = "cursor.cursor-agent",
                toolWindowId = "__terminal__:$id",
                isTerminalAgent = true,
            ),
        )
    }
}
